### PR TITLE
security added

### PR DESCRIPTION
**Pull request by AndreyDMA**

### CODE REVIEW

File: src/main/java/mate/academy/spring/controller/RentController.java [hunk 1]
```diff
@@ -25,23 +29,40 @@
 
     @GetMapping("/rentBook")
     public String rentBook(@RequestParam("book_id") Long id, ModelMap model) {
-        model.put("book", libraryService.rentBook(userService.getById(USER_ID),
+        Optional<User> currentUser = getCurrentUser();
+        if (currentUser.isEmpty()) {
+            return "null";
```
Comment (boroda4436) [inline src/main/java/mate/academy/spring/controller/RentController.java]: @AndreyDMA return "null" is a very bad idea! Can we handle in in another way? For example, redirect to login?

File: src/main/java/mate/academy/spring/controller/RentController.java [hunk 2]
```diff
@@ -25,23 +29,40 @@
...
                 bookService.getById(id)));
         return "book/all";
     }
 
     @GetMapping("/returnBook")
     public String returnBook(@RequestParam("book_id") Long id) {
-        Rent rent = libraryService.getRent(userService.getById(USER_ID),
+        Optional<User> currentUser = getCurrentUser();
+        if (currentUser.isEmpty()) {
+            return "null";
```
Comment (boroda4436) [inline src/main/java/mate/academy/spring/controller/RentController.java]: .

File: src/main/java/mate/academy/spring/controller/RentController.java [hunk 3]
```diff
@@ -25,23 +29,40 @@
...
                 bookService.getById(id));
         libraryService.returnBook(rent);
         return "book/all";
     }
 
     @GetMapping("/rentedBooks")
     public String getBooksRentedByUser(@RequestParam("user_id") Long id, ModelMap model) {
+        Optional<User> currentUser = getCurrentUser();
+        if (currentUser.isEmpty()) {
+            return "null";
```
Comment (boroda4436) [inline src/main/java/mate/academy/spring/controller/RentController.java]: .

File: src/main/java/mate/academy/spring/security/EmailValidator.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    private Pattern pattern;
+    private Matcher matcher;
+    private static final String EMAIL_PATTERN = "^[A-Z0-9._%+-]+@[A-Z0-9.-]+\\.[A-Z]{2,6}$";
+
+    @Override
+    public void initialize(ValidEmail constraintAnnotation) {
+    }
+
+    @Override
+    public boolean isValid(String email, ConstraintValidatorContext validatorContext) {
```
Comment (boroda4436) [inline src/main/java/mate/academy/spring/security/EmailValidator.java]: @AndreyDMA can we cover this with JUnit tests in the future?

File: src/main/java/mate/academy/spring/security/PasswordMatchesValidator.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package mate.academy.spring.security;
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+import mate.academy.spring.annotations.PasswordMatches;
+import mate.academy.spring.dto.UserDto;
+
+public class PasswordMatchesValidator implements ConstraintValidator<PasswordMatches, Object> {
```
Comment (boroda4436) [inline src/main/java/mate/academy/spring/security/PasswordMatchesValidator.java]:
```suggestion
public class PasswordMatchesValidator implements ConstraintValidator<PasswordMatches, UserDto> {
```

@AndreyDMA Can we do this in this way? Also, can we cover this with JUnit tests in the future?

File: src/main/java/mate/academy/spring/service/UserService.java [hunk 6]
```diff
@@ -1,14 +1,19 @@
...
-    void add(User user);
+    User add(User user);
 
     User getById(Long id);
 
     List<User> listUsers();
+
+    Optional<User> findByUsername(String username);
+
+    Optional<User> getByUsername(String username);
```
Comment (boroda4436) [inline src/main/java/mate/academy/spring/service/UserService.java]: @AndreyDMA what is the difference between this method and _ findByUsername_ ?

File: src/main/java/mate/academy/spring/controller/RegistrationController.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    PasswordEncoder passwordEncoder;
+
+    @GetMapping
+    public String regJsp() {
+        return "registration";
+    }
+
+    @PostMapping("/add")
+    public String registration(@ModelAttribute @Valid UserDto userDto, ModelMap model) {
+        User user = new User(userDto.getUsername(), passwordEncoder.encode(userDto.getPassword()),
```
[the remark this answers is not in the document]
Comment (AndreyDMA) [reply src/main/java/mate/academy/spring/controller/RegistrationController.java]: done

File: src/main/java/mate/academy/spring/controller/RegistrationController.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+                userDto.getName(), userDto.getSurname(), userDto.getEmail());
+        model.put("book", userService.add(user));
+        return "null";
```
[the remark this answers is not in the document]
Comment (AndreyDMA) [reply src/main/java/mate/academy/spring/controller/RegistrationController.java]: fixed

File: src/main/java/mate/academy/spring/controller/RentController.java (hunk 1 above)
Comment (AndreyDMA) [reply src/main/java/mate/academy/spring/controller/RentController.java]: fixed

File: src/main/java/mate/academy/spring/controller/RentController.java (hunk 2 above)
Comment (AndreyDMA) [reply src/main/java/mate/academy/spring/controller/RentController.java]: fixed

File: src/main/java/mate/academy/spring/controller/RentController.java (hunk 3 above)
Comment (AndreyDMA) [reply src/main/java/mate/academy/spring/controller/RentController.java]: fixed

File: src/main/java/mate/academy/spring/security/EmailValidator.java (hunk 4 above)
Comment (AndreyDMA) [reply src/main/java/mate/academy/spring/security/EmailValidator.java]: I think yes, but I'm not sure))

File: src/main/java/mate/academy/spring/security/PasswordMatchesValidator.java (hunk 5 above)
Comment (AndreyDMA) [reply src/main/java/mate/academy/spring/security/PasswordMatchesValidator.java]: I can't do it in this way, becouse I can't change signature of interface...

Comment (boroda4436) [reply src/main/java/mate/academy/spring/security/PasswordMatchesValidator.java]: @AndreyDMA 
https://github.com/TatianaPirch/jv-spring-hibernate/blob/master/src/main/java/mate/academy/spring/security/PasswordMatchesValidator.java#L10

Comment (boroda4436) [reply src/main/java/mate/academy/spring/security/PasswordMatchesValidator.java]: please fix and merge

Comment (boroda4436) [reply src/main/java/mate/academy/spring/security/PasswordMatchesValidator.java]: if you have a questions - ping me in slack

**Pull request merged**